lib/HLSL/DxilTargetTransformInfo.cpp

```diff
@@ -0,0 +1,96 @@
+//===-- DxilTargetTransformInfo.cpp - DXIL specific TTI pass     ----------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// \file
+// This file implements a TargetTransformInfo analysis pass specific to the
+// DXIL. Only implemented isSourceOfDivergence for DivergenceAnalysis.
+//
+//===----------------------------------------------------------------------===//
+
+#include "DxilTargetTransformInfo.h"
+#include "dxc/HLSL/DxilModule.h"
+#include "dxc/HLSL/DxilOperations.h"
+#include "llvm/CodeGen/BasicTTIImpl.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+#define DEBUG_TYPE "DXILtti"
+
+// For BasicTTImpl
+cl::opt<unsigned>
+    llvm::PartialUnrollingThreshold("partial-unrolling-threshold", cl::init(0),
+                                    cl::desc("Threshold for partial unrolling"),
+                                    cl::Hidden);
+
+DxilTTIImpl::DxilTTIImpl(const TargetMachine *TM, const Function &F,
+                         hlsl::DxilModule &DM, bool ThreadGroup)
+    : BaseT(TM, F.getParent()->getDataLayout()), m_pHlslOP(DM.GetOP()),
+      m_isThreadGroup(ThreadGroup) {}
+
+namespace {
+bool IsDxilOpSourceOfDivergence(const CallInst *CI, OP *hlslOP,
+                                bool ThreadGroup) {
+
+  DXIL::OpCode opcode = hlslOP->GetDxilOpFuncCallInst(CI);
+  switch (opcode) {
```

Comment:
We should auto-generate this from the instruction database (hctdb.py). We probably need to extend the instruction representation to say if an instruction is a source of divergence. Default value can be false.